### ISSUE TITLE
Full data mapper alters loaded data with pmap

### ISSUE THREAD

**Issue by S-Thaler** [tummfm/jax-sgmc#28]
I've encountered this strange problem: If the body function includes a pmap, the data loaded by the full_data_mapper seems to be mixed up and partially duplicated. The error only occurs if the code is run on more than 1 GPU. Interestingly, the initially cached batches are correct and also the shuffled random_data_loader works. Could it be some interference between pmap and data loading via host_callback?

```python
import os
os.environ["CUDA_VISIBLE_DEVICES"] = str('2,3')  # works if only 1 GPU selected

from jax import numpy as jnp, pmap, device_count
import numpy as onp

from jax_sgmc import data
from jax_sgmc.data import numpy_loader

dataset = onp.arange(100)
loader = numpy_loader.NumpyDataLoader(copy=False, data=dataset)

batch_size = 4
batch_cache = 8   # initial cashed batches are correct
n_devices = device_count()
batch_per_device = batch_size // n_devices
map_fn = data.full_data_mapper(loader, batch_cache, batch_size)


@pmap
def identity_fn(input):
    return input


def to_map_fn(batch, carry):
    input_copy = jnp.array(batch['data'], copy=True)
    pmap_batch = jnp.reshape(input_copy, (n_devices, batch_per_device))
    pmap_batch = identity_fn(pmap_batch)  # works if we don't run this line
    result = jnp.reshape(pmap_batch, -1)
    return (batch['data'], result), carry


(unchanged_data, identity_data), carry = map_fn(to_map_fn, 0.)
print('Output size:', identity_data.size, '\n', identity_data, unchanged_data)

# single batches work as expected:
init_batch_state, get_batch = data.random_reference_data(loader, batch_cache,
                                                         batch_size)
batch_state = init_batch_state(shuffle=True)
for i in range(300):
    batch_state, batch = get_batch(batch_state)
    (direct_identity, pmap_identity), _ = to_map_fn(batch, 0)
    print(f'Batch {i}: Error = {jnp.sum(jnp.abs(pmap_identity - batch["data"]))}')
```
